worker/upload.go

```diff
@@ -0,0 +1,1126 @@
+	data := make(chan slabData)
+
+	go func() {
+		for {
+			select {
+			case <-u.mgr.stopChan:
+				data <- slabData{err: errors.New("manager was stopped")}
+				return
+			case <-ctx.Done():
+				data <- slabData{err: errors.New("upload timed out")}
+				return
+			case <-u.nextReadTrigger:
+			}
+
+			buf := make([]byte, size)
+			length, err := io.ReadFull(io.LimitReader(r, size), buf)
+			if err == io.EOF {
+				close(data)
+				return
+			} else if err != nil && err != io.ErrUnexpectedEOF {
+				data <- slabData{err: errors.New("data read failed")}
+				return
+			}
+
+			data <- slabData{data: buf, length: length}
+		}
+	}()
+
+	return data
+}
+
+func (u *upload) triggerDoneShard() {
+	select {
+	case u.doneShardTrigger <- struct{}{}:
+	default:
+	}
+}
+
+func (u *upload) triggerNextRead() {
+	select {
+	case u.nextReadTrigger <- struct{}{}:
+	default:
+	}
+}
+
+func (u *upload) uploadShards(ctx context.Context, shards [][]byte, index int) ([]object.Sector, error) {
+	ctx, span := tracing.Tracer.Start(ctx, "uploadShards")
+	defer span.End()
+
+	// convenience variables
+	mgr := u.mgr
+
+	// prepare the upload
+	slab, uploads, shardRespChan := u.newSlabUpload(ctx, shards)
+	span.SetAttributes(attribute.Stringer("id", slab.sID))
+	defer u.finishSlabUpload(slab)
+
+	// launch all shard uploads
+	for _, upload := range uploads {
+		if err := slab.launch(upload); err != nil {
+			return nil, err
+		}
+	}
+
+	// create a timer to trigger overdrive
+	timeout := time.NewTimer(mgr.overdriveTimeout)
+	resetTimeout := func() {
+		timeout.Stop()
+		select {
+		case <-timeout.C:
+		default:
+		}
+		timeout.Reset(mgr.overdriveTimeout)
+	}
+
+	// launch a goroutine to trigger overdrive
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timeout.C:
+				if upload := slab.overdrive(shardRespChan, shards); upload != nil {
+					_ = slab.launch(upload) // ignore error
+				}
+				resetTimeout()
+			}
+		}
+	}()
+
+	// collect responses
+	var finished bool
+	for slab.inflight() > 0 && !finished {
+		var resp shardResp
+		select {
+		case <-mgr.stopChan:
+			return nil, errors.New("upload stopped")
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		case resp = <-shardRespChan:
+		}
+
+		// receive the response
+		finished = slab.receive(resp)
+
+		// handle the response
+		if resp.err == nil {
+			u.triggerDoneShard()
+			if slab.shouldTriggerNextRead() {
+				u.triggerNextRead()
+			}
+
+			resetTimeout()
+		}
+
+		// relaunch non-overdrive uploads
+		if resp.err != nil && !resp.req.overdrive {
+			if err := slab.launch(resp.req); err != nil {
+				break // fail the download
+			}
+		}
+	}
+
+	// register the amount of overdrive sectors
+	span.SetAttributes(attribute.Int("overdrive", slab.overdriveCnt()))
+
+	// track stats
+	mgr.statsOverdrive.track(slab.overdrivePct())
+	mgr.statsSpeed.track(float64(slab.bytesPerMS()))
+	return slab.finish()
+}
+
+func (u *uploader) start(hp hostProvider, rl revisionLocker) {
+	go func() {
```

Comment:
If the whole method is a go func it might be worth to launch this with `go start()` instead.